core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java

```diff
@@ -0,0 +1,104 @@
+package org.activityinfo.model.type.subform;
+/*
+ * #%L
+ * ActivityInfo Server
+ * %%
+ * Copyright (C) 2009 - 2013 UNICEF
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the 
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public 
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+import com.google.common.collect.Maps;
+import org.activityinfo.model.form.FormClass;
+import org.activityinfo.model.legacy.CuidAdapter;
+import org.activityinfo.model.resource.ResourceId;
+
+import java.util.Map;
+
+/**
+ * Represents type of form classes : locationType classes or activity classes.
+ * Mainly used to allow user define subform kind.
+ *
+ * @author yuriyz on 03/12/2015.
+ */
+public enum ClassType {
```

Comment:
But why handle Partner, Project, Location FormClasses specially? Why wouldn't we just address the general case of "one subform for each resource in a given FormClass" ?

And taking this a step further, whether they're periodic forms or bound to a collection, what they all have in common is a "key" field. For monthly forms it would be a key of type MonthType, if you wanted one entry for each location, the key field would be of type ReferenceType. This would make it easy to expand this to a text field for user-supplied IDs, etc. 

Then basically the only additions we would need to make to the model would be:

```.java
FormClass {
  /* add: */
  Optional<ResourceId> parentFormId;   /* added */
  Optional<FormField> keyField;             /* added */
}
SubFormType : FieldType {
  ResourceId subFormId;
}
```

For SubForm Resources, we could use two reserved words to store the key and the parent id: for example, `__key` and `__parent`

Do you think this would work? 

It would give us a pretty uniform path to dealing with one of the most challenging aspects of the subforms: maintaining the key constraint. We will need to deal with the case of two users submitting a "sub resource" for the same parent and the same key. This gives us a common way of handling all of the cases.